File: iOS/ChatRelated/ViewModels/ChatViewModel.swift
//
//  ChatViewModel.swift
//  SwiftiDate
//
//  Created by 游哲維 on 2025/3/6.
//

import Foundation
import SwiftUI
import Firebase

class ChatViewModel: ObservableObject {
    // MARK: - Published State
    @Published var selectedChat: Chat? = nil
    @Published var userMatches: [UserMatch] = []
    @Published var chatData: [Chat] = []
    @Published var interactiveMessage: [Message] = []
    @Published var chatMessages: [UUID: [Message]] = [:]
    @Published var showInteractiveContent: Bool = false
    @Published var showTurboPurchaseView: Bool = false
    @Published var showTurboView: Bool = false
    @Published var selectedTurboTab: Int = 0
    @Published var showSafetyCenterView: Bool = false
    @Published var showSearchField: Bool = false
    @Published var searchText: String = ""
    
    // MARK: - AppStorage Keys
    @AppStorage("userMatchesStorage") var userMatchesString: String = ""
    @AppStorage("chatDataStorage") var chatDataString: String = ""
    @AppStorage("chatMessagesStorage") var chatMessagesString: String = ""
    
    // MARK: - Computed Properties for Filtering
    var filteredMatches: [UserMatch] {
        if searchText.isEmpty { return userMatches }
        return userMatches.filter { $0.name.contains(searchText) }
    }
    
    var filteredChats: [Chat] {
        if searchText.isEmpty { return chatData }
        return chatData.filter { $0.name.contains(searchText) }
    }
    
    // MARK: - Data Loading Methods
    func loadChats() {
        readDataFromFirebase()
    }
    
    private let userSettings: UserSettings
    
    init(userSettings: UserSettings) {
        self.userSettings = userSettings
    }
    
    func readDataFromFirebase() {
        let ref = Database.database(url: "https://swiftidate-cdff0-default-rtdb.asia-southeast1.firebasedatabase.app").reference()
        let userId = userSettings.globalUserID
        print("userSettings.globalUserID: \(userId)")

        // 讀取 userMatches
        ref.child("users").child(userId).child("userMatches").observeSingleEvent(of: .value) { snapshot in
            guard let value = snapshot.value as? [[String: Any]] else {
                print("Failed to decode userMatches data")
                return
            }
            
            do {
                let jsonData = try JSONSerialization.data(withJSONObject: value, options: [])
                print("userMatches json data: \(jsonData)")
                var userMatches = try JSONDecoder().decode([UserMatch].self, from: jsonData)
                
                // 將 userMatches 倒序排序
                userMatches.reverse()
                
                // 更新 self.userMatches 並存儲到本地
                self.userMatches = userMatches
                self.saveUserMatchesToAppStorage()
            } catch {
                // 这里使用 try? 防止在 catch 块中再次抛出错误
                if let jsonData = try? JSONSerialization.data(withJSONObject: value, options: []) {
                    print("userMatches json data: \(jsonData)")
                }
                print("Failed to decode userMatches: \(error)")
            }
        }

        // 讀取 chatData
        ref.child("users").child(userId).child("chats").observeSingleEvent(of: .value) { snapshot in
            guard let value = snapshot.value as? [[String: Any]] else {
                print("Failed to decode chats data")
                return
            }
            
            do {
                let jsonData = try JSONSerialization.data(withJSONObject: value, options: [])
                var chatData = try JSONDecoder().decode([Chat].self, from: jsonData)

                // 確保 chatData 至少有兩個元素
                if chatData.count > 1 {
                    // 保留第一個元素
                    let firstChat = [chatData[0]]
                    // 倒序排列剩下的元素
                    let reversedChats = chatData[1...].reversed()
                    // 合併結果
                    chatData = firstChat + reversedChats
                }

                self.chatData = chatData
                self.saveChatDataToAppStorage()
            } catch {
                print("Failed to decode chats: \(error)")
            }
        }
        
        print("Path: \(ref.child("users").child(userId).child("chatMessages").description())")
        
        // 讀取 chatMessages
        ref.child("users").child(userId).child("chatMessages").observeSingleEvent(of: .value) { snapshot in
            if snapshot.exists() {
                print("Snapshot exists for chatMessages: \(snapshot.value ?? "nil")")
            } else {
                print("Snapshot does not exist for chatMessages")
            }
            guard let value = snapshot.value as? [String: [[String: Any]]] else {
                print("Failed to decode chatMessages data")
                return
            }
            
            var chatMessages: [UUID: [Message]] = [:]
            do {
                for (key, messagesArray) in value {
                    guard let chatId = UUID(uuidString: key) else {
                        print("Invalid UUID: \(key)")
                        continue
                    }
                    
                    print("Processing chat ID: \(chatId)")
                    let jsonData = try JSONSerialization.data(withJSONObject: messagesArray, options: [])
                    if let jsonString = String(data: jsonData, encoding: .utf8) {
                        print("Serialized JSON for chat \(chatId): \(jsonString)")
                    }
                    
                    do {
                        let messages = try JSONDecoder().decode([Message].self, from: jsonData)
//                        print("Decoded Messages for chat \(chatId): \(messages)")
                        chatMessages[chatId] = messages
                    } catch let DecodingError.dataCorrupted(context) {
                        print("Data corrupted: \(context)")
                    } catch let DecodingError.keyNotFound(key, context) {
                        print("Key '\(key)' not found: \(context.debugDescription)")
                        print("Coding Path: \(context.codingPath)")
                    } catch let DecodingError.typeMismatch(type, context) {
                        print("Type '\(type)' mismatch: \(context.debugDescription)")
                        print("Coding Path: \(context.codingPath)")
                    } catch let DecodingError.valueNotFound(value, context) {
                        print("Value '\(value)' not found: \(context.debugDescription)")
                        print("Coding Path: \(context.codingPath)")
                    } catch {
                        print("Failed to decode Messages for chat \(chatId): \(error)")
                    }
                }
                
                // Update the state on the main thread
                DispatchQueue.main.async {
                    self.chatMessages = chatMessages
                    print("Final chatMessages dictionary: \(chatMessages)")
                    self.saveChatMessagesToAppStorage()
                }
                
            } catch {
                print("Failed to decode chatMessages: \(error)")
            }
        }
    }

    // MARK: - AppStorage Save/Load Methods
    func saveUserMatchesToAppStorage() {
        do {
            let encoder = JSONEncoder()
            let data = try encoder.encode(userMatches)
            userMatchesString = String(data: data, encoding: .utf8) ?? ""
        } catch {
            print("Failed to encode userMatches: \(error)")
        }
    }
    
    func loadUserMatchesFromAppStorage() {
        guard !userMatchesString.isEmpty else {
            print("No user matches found in AppStorage")
            return
        }
        
        do {
            let decoder = JSONDecoder()
            if let data = userMatchesString.data(using: .utf8) {
                userMatches = try decoder.decode([UserMatch].self, from: data)
                // 埋點：成功從本地載入 userMatches
                AnalyticsManager.shared.trackEvent("user_matches_loaded", parameters: [
                    "count": userMatches.count
                ])
            }
        } catch {
            print("Failed to decode userMatches: \(error)")
        }
    }
    
    func saveChatDataToAppStorage() {
        do {
            let encoder = JSONEncoder()
            let data = try encoder.encode(chatData)
            chatDataString = String(data: data, encoding: .utf8) ?? ""
        } catch {
            print("Failed to encode chatData: \(error)")
        }
    }
    
    func loadChatDataFromAppStorage() {
        guard !chatDataString.isEmpty else {
            print("No chat data found in AppStorage")
            return
        }
        
        do {
            let decoder = JSONDecoder()
            if let data = chatDataString.data(using: .utf8) {
                chatData = try decoder.decode([Chat].self, from: data)
                AnalyticsManager.shared.trackEvent("chat_data_loaded", parameters: [
                    "chat_count": chatData.count
                ])
            }
        } catch {
            print("Failed to decode chatData: \(error)")
        }
    }
    
    func saveChatMessagesToAppStorage() {
        do {
            let encoder = JSONEncoder()
            let data = try encoder.encode(chatMessages)
            chatMessagesString = String(data: data, encoding: .utf8) ?? ""
        } catch {
            print("Failed to encode chatMessages: \(error)")
        }
    }
    
    func loadChatMessagesFromAppStorage() {
        guard !chatMessagesString.isEmpty else {
            print("No chat messages found in AppStorage")
            return
        }
        
        do {
            let decoder = JSONDecoder()
            if let data = chatMessagesString.data(using: .utf8) {
                // 验证 JSON 数据是否有效
                if let jsonObject = try? JSONSerialization.jsonObject(with: data, options: []) {
                    print("Valid JSON: \(jsonObject)")
                } else {
                    print("Invalid JSON format")
                    return
                }
                
                // 尝试解码为 [UUID: [Message]]
                chatMessages = try decoder.decode([UUID: [Message]].self, from: data)
                print("Decoded chatMessages: \(chatMessages)")
                AnalyticsManager.shared.trackEvent("chat_messages_loaded", parameters: [
                    "chats_loaded": chatMessages.count
                ])
            }
        } catch {
            print("Failed to decode chatMessages: \(error)")
        }
    }
    
    // 方便更新聊天訊息
    func updateChatMessages(for chatID: UUID, messages: [Message]) {
        chatMessages[chatID] = messages
        saveChatMessagesToAppStorage() // 保存至 AppStorage
        AnalyticsManager.shared.trackEvent("chat_messages_updated", parameters: [
            "chat_id": chatID.uuidString,
            "message_count": messages.count
        ])
    }
    
    // For checking if local storage is empty
    var chatDataStringIsEmptyOrChatMessagesStringIsEmpty: Bool {
        return chatDataString.isEmpty || chatMessagesString.isEmpty
    }
}

File: flutter/swiftidate/lib/view_models/chat_view_model.dart
import 'dart:convert';
import 'package:firebase_core/firebase_core.dart';
import 'package:firebase_database/firebase_database.dart';
import 'package:flutter/material.dart'; // 新增這行
import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';
import '../models/chat.dart';
import '../models/message.dart';
import '../providers/user_settings.dart';

class UserMatch {
  final String name;
  final String imageName; // 新增這個屬性

  UserMatch({required this.name, required this.imageName});

  factory UserMatch.fromJson(Map<String, dynamic> json) {
    return UserMatch(
      name: json['name'] ?? '',
      imageName: json['imageName'] ?? 'assets/default_avatar.png', // 提供預設圖片路徑
    );
  }

  // 添加 toJson 方法以便序列化
  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'imageName': imageName,
    };
  }
}

class ChatViewModel extends ChangeNotifier {
  // 新增搜尋控制器
  final TextEditingController searchController = TextEditingController();

  // 主要狀態
  Chat? selectedChat;
  List<UserMatch> userMatches = [];
  List<Chat> chatData = [];
  List<Message> interactiveMessage = [];
  Map<String, List<Message>> chatMessages = {};
  bool showInteractiveContent = false;
  bool showTurboPurchaseView = false;
  bool showTurboView = false;
  int selectedTurboTab = 0;
  bool showSafetyCenterView = false;
  bool showSearchField = false;
  String searchText = '';

  // 使用 SharedPreferences 模擬 AppStorage 的行為
  String userMatchesString = '';
  String chatDataString = '';
  String chatMessagesString = '';

  // 計算過濾後的清單
  List<UserMatch> get filteredMatches {
    if (searchText.isEmpty) return userMatches;
    return userMatches.where((match) => match.name.contains(searchText)).toList();
  }

  List<Chat> get filteredChats {
    if (searchText.isEmpty) return chatData;
    return chatData.where((chat) => chat.name.contains(searchText)).toList();
  }

  final UserSettings userSettings;

  ChatViewModel({required this.userSettings});

  void loadChats() {
    readDataFromFirebase();
  }

  Future<void> readDataFromFirebase() async {
    // 初始化 FirebaseApp，如果尚未初始化
    await Firebase.initializeApp();
    final ref = FirebaseDatabase.instanceFor(
      app: Firebase.app(),
      databaseURL: "https://swiftidate-cdff0-default-rtdb.asia-southeast1.firebasedatabase.app",
    ).reference();
    final userId = userSettings.globalUserID;
    print("userSettings.globalUserID: $userId");

    // 讀取 userMatches
    DatabaseEvent eventMatches =
        await ref.child("users").child(userId).child("userMatches").once();
    final valueMatches = eventMatches.snapshot.value;
    if (valueMatches is List) {
      try {
        final jsonData = jsonEncode(valueMatches);
        List<UserMatch> matches = (jsonDecode(jsonData) as List)
            .map((item) => UserMatch.fromJson(item))
            .toList();
        // 倒序排列
        matches = matches.reversed.toList();
        userMatches = matches;
        await saveUserMatchesToAppStorage();
        notifyListeners();
      } catch (e) {
        print("Failed to decode userMatches: $e");
      }
    } else {
      print("Failed to decode userMatches data");
    }

    // 讀取 chatData
    DatabaseEvent eventChats =
        await ref.child("users").child(userId).child("chats").once();
    final valueChats = eventMatches.snapshot.value;
    if (valueChats is List) {
      try {
        final jsonData = jsonEncode(valueChats);
        List<Chat> chats = (jsonDecode(jsonData) as List)
            .map((item) => Chat.fromJson(item))
            .toList();
        if (chats.length > 1) {
          final firstChat = [chats[0]];
          final reversedChats = chats.sublist(1).reversed.toList();
          chats = firstChat + reversedChats;
        }
        chatData = chats;
        await saveChatDataToAppStorage();
        notifyListeners();
      } catch (e) {
        print("Failed to decode chats: $e");
      }
    } else {
      print("Failed to decode chats data");
    }

    // 讀取 chatMessages
    DatabaseEvent eventChatMessages =
        await ref.child("users").child(userId).child("chatMessages").once();
    final valueChatMessages = eventChatMessages.snapshot.value;
    if (valueChatMessages is Map) {
      Map<String, List<Message>> tempChatMessages = {};
      try {
        valueChatMessages.forEach((key, messagesArray) {
          if (messagesArray is List) {
            final jsonData = jsonEncode(messagesArray);
            try {
              List<Message> messages = (jsonDecode(jsonData) as List)
                  .map((item) => Message.fromJson(item))
                  .toList();
              tempChatMessages[key] = messages;
            } catch (e) {
              print("Failed to decode Messages for chat $key: $e");
            }
          }
        });
        chatMessages = tempChatMessages;
        await saveChatMessagesToAppStorage();
        notifyListeners();
      } catch (e) {
        print("Failed to decode chatMessages: $e");
      }
    } else {
      print("Failed to decode chatMessages data");
    }
  }

  // 以下方法使用 SharedPreferences 作存儲示例
  Future<void> saveUserMatchesToAppStorage() async {
    try {
      final data = jsonEncode(userMatches.map((e) => e.toJson()).toList());
      SharedPreferences prefs = await SharedPreferences.getInstance();
      await prefs.setString('userMatchesStorage', data);
      userMatchesString = data;
    } catch (e) {
      print("Failed to encode userMatches: $e");
    }
  }

  Future<void> loadUserMatchesFromAppStorage() async {
    SharedPreferences prefs = await SharedPreferences.getInstance();
    userMatchesString = prefs.getString('userMatchesStorage') ?? '';
    if (userMatchesString.isNotEmpty) {
      try {
        final List decoded = jsonDecode(userMatchesString);
        userMatches = decoded.map((e) => UserMatch.fromJson(e)).toList();
        notifyListeners();
      } catch (e) {
        print("Failed to decode userMatches: $e");
      }
    }
  }

  Future<void> saveChatDataToAppStorage() async {
    try {
      final data = jsonEncode(chatData.map((e) => e.toJson()).toList());
      SharedPreferences prefs = await SharedPreferences.getInstance();
      await prefs.setString('chatDataStorage', data);
      chatDataString = data;
    } catch (e) {
      print("Failed to encode chatData: $e");
    }
  }

  Future<void> loadChatDataFromAppStorage() async {
    SharedPreferences prefs = await SharedPreferences.getInstance();
    chatDataString = prefs.getString('chatDataStorage') ?? '';
    if (chatDataString.isNotEmpty) {
      try {
        final List decoded = jsonDecode(chatDataString);
        chatData = decoded.map((e) => Chat.fromJson(e)).toList();
        notifyListeners();
      } catch (e) {
        print("Failed to decode chatData: $e");
      }
    }
  }

  Future<void> saveChatMessagesToAppStorage() async {
    try {
      // 將 Map 序列化，這裡假設 key 為 String
      final data = jsonEncode(chatMessages.map((key, value) =>
          MapEntry(key, value.map((e) => e.toJson()).toList())));
      SharedPreferences prefs = await SharedPreferences.getInstance();
      await prefs.setString('chatMessagesStorage', data);
      chatMessagesString = data;
    } catch (e) {
      print("Failed to encode chatMessages: $e");
    }
  }

  Future<void> loadChatMessagesFromAppStorage() async {
    SharedPreferences prefs = await SharedPreferences.getInstance();
    chatMessagesString = prefs.getString('chatMessagesStorage') ?? '';
    if (chatMessagesString.isNotEmpty) {
      try {
        final Map decoded = jsonDecode(chatMessagesString);
        chatMessages = decoded.map((key, value) => MapEntry(
            key.toString(),
            (value as List)
                .map((item) => Message.fromJson(item))
                .toList()));
        notifyListeners();
      } catch (e) {
        print("Failed to decode chatMessages: $e");
      }
    }
  }

  // 更新聊天訊息的輔助方法
  void updateChatMessages(String chatID, List<Message> messages) {
    chatMessages[chatID] = messages;
    saveChatMessagesToAppStorage();
    // 此處可加入 Analytics 追蹤
    notifyListeners();
  }

  bool get isLocalStorageEmpty {
    return chatDataString.isEmpty || chatMessagesString.isEmpty;
  }
}
